mock/py/mockbuild/external.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+# vim:expandtab:autoindent:tabstop=4:shiftwidth=4:filetype=python:textwidth=0:
+
+from .exception import ExternalDepsError
+from .trace_decorator import traceLog
+
+
+class ExternalDeps(object):
+    """ Handles external dependecies. E.g external:* """
+
+    def __init__(self, buildroot, bootstrap_buildroot, uid_manager):
+        self.buildroot = buildroot
+        self.bootstrap_buildroot = bootstrap_buildroot
+        self.uid_manager = uid_manager
+
+    # when python 3.9 becomes standard, this can be replaced by string.removeprefix()
+    @classmethod
+    def _remove_prefix(cls, intext, prefix):
+        if intext.startswith(prefix):
+            return intext[len(prefix):]
+        return intext
+
+    def extract_external_deps(self, requires):
+        """ accepts list of (build)requires, returns all external deps """
+        return [i for i in requires if i.startswith('external:')]
+
+    @traceLog()
+    def install_external_deps(self, deps):
+        """Install dependencies using native library manager"""
+        self.buildroot.root_log.info('Installing dependencies to satisfy external:*')
+
+        prefix = 'external:pypi:'
+        pypi_deps = [self._remove_prefix(i, prefix) for i in deps if i.startswith(prefix)]
+        deps = [i for i in deps if not i.startswith(prefix)]
+        if pypi_deps:
+            self.install_external_deps_pypi(pypi_deps)
+
+        prefix = 'external:crate:'
+        crate_deps = [self._remove_prefix(i, prefix) for i in deps if i.startswith(prefix)]
+        deps = [i for i in deps if not i.startswith(prefix)]
+        if crate_deps:
+            self.install_external_deps_crate(crate_deps)
+
+        if deps:
+            raise ExternalDepsError("Unknown external dependencies: {}".format(', '.join(deps)))
+
+    @traceLog()
+    def install_external_deps_pypi(self, deps):
+        """ deps is list of python modules. Without the prefix. """
+        self.bootstrap_buildroot.install_as_root('/usr/bin/pip3', 'python3-setuptools')
```

Comment:
When bootstrap chroot is disabled, we need to stop building much sooner then here.